biothings/utils/sqlite3.py

```diff
@@ -335,27 +336,65 @@ def find(self, *args, **kwargs):
         else:
             raise NotImplementedError("find: args=%s kwargs=%s" % (repr(args), repr(kwargs)))
 
-    def insert_one(self, doc):
-        assert "_id" in doc
-        with self.get_conn() as conn:
-            conn.execute(
-                "INSERT INTO %s (_id,document) VALUES (?,?)" % self.colname,
-                (doc["_id"], json.dumps(doc, default=json_serial)),
-            ).fetchone()
-            conn.commit()
+    def insert_one(self, doc: dict) -> None:
+        """
+        single-document insert into the database
 
-    def insert(self, docs, *args, **kwargs):
+        Leverages the execute function using the question mark style for specifying
+        the values to insert into the table
+        https://docs.python.org/3/library/sqlite3.html#sqlite3.Connection.execute
+        https://docs.python.org/3/library/sqlite3.html#how-to-use-placeholders-to-bind-values-in-sql-queries
+        """
         with self.get_conn() as conn:
-            for doc in docs:
-                conn.execute(
-                    "INSERT INTO %s (_id,document) VALUES (?,?)" % self.colname,
-                    (doc["_id"], json.dumps(doc, default=json_serial)),
-                ).fetchone()
-                conn.commit()
+            try:
+                parameters = ((doc["_id"], json.dumps(doc, default=json_serial)),)
+                sql_statement = f"INSERT INTO {self.colname} (_id,document) VALUES (?,?)"
+                conn.execute(sql_statement, parameters)
+            except sqlite3.IntegrityError as integrity_err:
+                logger.exception(integrity_err)
+                logger.error("Unable to complete transation (check the _id value for uniqueness). Document: %s", doc)
+                raise integrity_err
+
+    def insert(self, docs: list[dict]) -> None:
```

Comment:
here can we still keep the option of allowing `docs` as an iterator? Useful for a large input

Reply:
@newgene So we can pass in an iterator to the method call, but in order to get the speed gains we'll have to generate some structure for sqlite to consume. If we have a low-memory constraint we wish to maintain, I'd likely want to create a separate API / method for it because we'll ultimately just generate the structure in this one when we call `executemany`